### ISSUE TITLE
dart js-interop errors and execptions

### ISSUE THREAD

**Issue by kealjones-wk** [dart-lang/sdk#36363]
Setup: 
* Dart SDK Version: 2.2.0
* MacOSX
* Chrome

While working on `cleandart/react-dart` we came across an issue that I am hoping to get some clarification on.

When a dart exception is thrown and a javascript callback passes this error into dart the error comes through as a `JsObject` in dart2js and as a `NativeJavaScriptObject` in DDC. Is this an issue or is this expected behavior?

We have found ways around this, that i wanted ask about and see if they are safe ways to go about re-dartifying thrown exceptions or if these solutions were potentially problematic:

in dart2js: 
`_JoesException joesError = js_util.getProperty(error, 'dartException');`

in ddc:
`_JoesException joesError = _getSymbolProperty(error, 'Symbol(_thrownValue)');`
this requires a js side function used through js interop:
```js
function _getSymbolProperty(obj, key){
  var sym = Object.getOwnPropertySymbols(obj).find(function(s) {
    return String(s) === key;
  });
  return obj[sym];
}
```

Any insight or better solutions would be welcome, thanks!

**Comment by kealjones-wk** [comment]
Any thoughts? Would a branch help? I could throw one togeatger that shows the issue we are hitting.

**Comment by aaronlademann-wf** [comment]
@kevmoo @robbecker-wf @johnpryan

**Comment by kealjones-wk** [comment]
Just hit this issue again for https://github.com/Workiva/react_testing_library/

**Comment by kevmoo** [comment]
@srujzs - is this on your radar?

**Comment by srujzs** [comment]
Nope, but now it is. :) The difference in type you see between the compilers is just a difference in the way the two compilers reify unknown JS objects. It might be worth unifying the runtime type so it's less confusing, but that's tangential.

It's interesting that the exception is caught when it's directly thrown from JS but not in the `dart:html` case. I'll have to repro the different examples locally and report back.

**Comment by KealJones** [comment]
@srujzs I hope this helps, I wrote a dartpad based example that demonstrates getting the error in dart then throwing in js to dartify:
Gist: https://gist.github.com/kealjones-wk/0e955521d53848aef580ebc837642b47
Dartpad: https://dartpad.dev/?id=0e955521d53848aef580ebc837642b47

**Comment by srujzs** [comment]
I think the behavior you see is a consequence of the way we handle exceptions in the web compilers. We wrap values that are thrown and then they are unwrapped when caught in a catch block. In the Dart world, this is fine, but it presents issues with interop. Since we never catch the error, the error never gets unwrapped, which is what you've noticed.

It's also the same reason why throwing it from JS and catching it in Dart ends up unwrapping the error. The compilers don't try to unwrap something that's already unwrapped, so even if you called `throwFromJS('hello world')` and catch it in Dart, it'd work as expected. The struggle though is that if we throw in Dart, we don't know whether or not it's caught in Dart or JS, so we don't know concretely whether it should be wrapped or not.

Fixing this might be complicated, but I suppose there are two options:

1) Ironically enough, rethrowing through JS and catching it in Dart is a solution. We could make this simpler by adding similar functionality to a JS library (or have it just unwrap if possible directly).
2) A potentially better solution might be to make some changes to the RTI to handle the wrapped errors. `is` and `as` would handle it such that the underlying error also gets checked, with the `as` check doing the unwrapping. Right now, the wrapped errors are `JavaScriptObject`s so that'll need to be changed. This would at least allow you to treat the error as the expected Dart value, even though it requires a cast (although maybe we can modify the error streams in `dart:html` to do these casts beforehand). I'm not sure what complications might arise from this however. For example, are there instances where we don't want wrapped errors to be treated as the underlying error?

@sigmundch, any thoughts on this?

**Comment by sigmundch** [comment]
/cc @rakudrama 

I too feel that maybe this is should have been recognized by our error streams in dart:html. Today it exposes the errors as they are, but we know they could have originated from Dart, in which case, it seems reasonable that the API would provide them transparently as Dart errors.

That said, generally we have too much magic in our dart:html APIs. Yet another option to consider here is whether we could expose the unwrapping somehow. For example, by allowing `Error` to be a static interop type, and expose a getter with the underlying Dart error. Maybe this is not an option for dart:html that intends to be batteries included, but that may be the option for how one using JS-interop to access browser event streams could manually unwrap when needed.

**Comment by kealjones-wk** [comment]
The original issue I posted doesn't use any dart:html streams, the error is caught in JS and then handed to an `allowInterop`d function by react. So, do like the idea of something clean like an `as` cast or a function in a dart JS library. Maybe similar to `jsify`, but in reverse, or `promiseToFuture`, so that we could pass the error in and get the unwrapped dart version back.

At least on this particular Workiva team we are not scared of the interop or having to do things manually. Personally I prefer `dart:js_util` over `dart:js` anyday. Meaning, we don't need magic, I just want some way to be able to unwrap these errors that I know originated from dart.
